contracts/GroupVoteLocker.sol

```diff
@@ -0,0 +1,331 @@
+        uint256 epoch;
+    }
+
+    /* @dev Update the group voting state
+     * @param oldLockup Users lockup prior to the change that triggered the checkpoint
+     * @param newLockup Users new lockup
+     * @param oldCheckpoint checkpoint related to oldLockup
+     * @param newCheckpoint checkpoint related to newLockup
+     * @param groupState group state to be modified
+     */
+    function _updateGroupState(
+        Lockup memory oldLockup,
+        Lockup memory newLockup,
+        Checkpoint memory oldCheckpoint,
+        Checkpoint memory newCheckpoint,
+        GroupVotePowerState storage groupState
+    )
+        internal 
+    {
+        int128 oldSlopeDelta = 0;
+        int128 newSlopeDelta = 0;
+
+        oldSlopeDelta = groupState.slopeChanges[oldLockup.end];
+        if (newLockup.end != 0) {
+            if (newLockup.end == oldLockup.end) {
+                // Lockup dates are the same end time, slope delta is the same
+                newSlopeDelta = oldSlopeDelta;
+            } else {
+                newSlopeDelta = groupState.slopeChanges[newLockup.end];
+            }
+        }
+
+        _writeGroupCheckpoint(
+            newCheckpoint.slope - oldCheckpoint.slope,
+            newCheckpoint.bias - oldCheckpoint.bias,
+            groupState
+        );
+
+        /* Schedule the slope changes. There is a possible code simplification where
+         * we always undo the old checkpoint slope change and always apply the new
+         * checkpoint slope change. In the interest of gas optimization the code is 
+         * slightly more complicated.
+         */
+
+        // old lockup still active and needs slope change adjustment. 
+        if (oldLockup.end > block.timestamp) {
+            // this is an adjustment of the slope: oldSlopeDelta was <something> - oldCheckpoint.slope, 
+            // so we cancel/undo that
+            oldSlopeDelta = oldSlopeDelta + oldCheckpoint.slope;
+            // gas optimize it so another storage access for newLockup is not required
+            if (newLockup.end == oldLockup.end) {
+                // It was a new deposit, not extension
+                oldSlopeDelta = oldSlopeDelta - newCheckpoint.slope;
+            }
+            groupState.slopeChanges[oldLockup.end] = oldSlopeDelta;
+        }
+        if (newLockup.end > block.timestamp) {
+            // (second part of gas optimization) it was an extension
+            if (newLockup.end > oldLockup.end) {
+                newSlopeDelta = newSlopeDelta - newCheckpoint.slope;
+                groupState.slopeChanges[newLockup.end] = newSlopeDelta;
+            }
+        }
+    }
+
+    /**
+     * @dev Add a checkpoint to the group state
+     * @param userSlopeDelta Change in slope that triggered this checkpoint
+     * @param userBiasDelta Change in bias that triggered this checkpoint
+     * @param groupState group state to have a checkpoint added
+     */
+    function _writeGroupCheckpoint(
```

Comment:
🔴 I've not thoroughly checked out to see if it's a bug, but it feels like it might be one. Since we always write at least one checkpoint when write is called, there can be more than one checkpoint for a single block.

When searching for a block's checkpoints, there is a deterministic algorithm that runs to find the checkpoint, however it does not guarantee that it will select the last checkpoint per block, and could select any of the block's checkpoints. Almost everything we do involves comparing values at a time of two different checkpoint lists against each other. An attacker could possibly manipulate this such that the the user and the total values for a block were from different checkpoint orders, and thus bad things could happen.

Reply:
Definitely something to investigate. Though I think `epochs` are used rather than solely block numbers to find checkpoints. And each time a new one is created an epoch increases as well.